### ISSUE TITLE
Python Azure and ServiceBus

### ISSUE THREAD

**Issue by riccardobecker** [Azure/azure-sdk-for-python#340]
Hi All,

I am fairly new to python but I have an issue with the ServiceBus wrapper. I have devices that will contain python code to listen to events in a specific subscription for that device. I don't want the devices to have access to the whole servicebus namespace (and use the RootManageSharedAccessKey). I already created a SAS token on the backend for every specific deviceid.

How can I get python to work with the generated SAS token as it can be used in e.g. c# with the TokenProvider class. See below for C#, I am looking for a python equivalient.

  MessagingFactorySettings mf = new MessagingFactorySettings();
            mf.TokenProvider = TokenProvider.CreateSharedAccessSignatureTokenProvider(_listenSAS);
            MessagingFactory fac = MessagingFactory.Create("sb://" + <namespace> + ".servicebus.windows.net", mf);
            SubscriptionClient subScriptionClient = fac.CreateSubscriptionClient(<TOPIC>,
                _deviceId);

regards

**Comment by huguesv** [comment]
Hi!

SAS support in Python works like this:

```
key_name = 'RootManageSharedAccessKey' # or any policy name you created in the portal configure page
key_value = '' # the primary or secondary key for that policy
sbs = ServiceBusService(service_namespace,
                        shared_access_key_name=key_name,
                        shared_access_key_value=key_value)
```

**Comment by riccardobecker** [comment]
Thanks Hugues,

This part I know and works. What i have is a sas generated with a specific expiry for every device that runs this python code. I dont want to give my credentials to all tge device, i just use my creds to generate a specific listen token for a specific subscription. Do this makes to you? I have the equivalent c# code running btw.

Sent from my Windows Phone
________________________________

**Comment by huguesv** [comment]
The ServiceBusService class accepts an `authentication` parameter which lets you customize the signing of requests.
https://github.com/Azure/azure-sdk-for-python/blob/master/azure/servicebus/servicebusservice.py#L71

If it's specified, it will call `sign_request` on that object.

If it's None, it will instantiate either the old `ServiceBusWrapTokenAuthentication` when `account_key` and `issuer` are given, or `ServiceBusSASAuthentication` for `shared_access_key_name` + ``shared_access_key_value` pair:
https://github.com/Azure/azure-sdk-for-python/blob/master/azure/servicebus/servicebusservice.py#L1185

You could create your own authentication class, inspired or derived from `ServiceBusSASAuthentication`, which will give you more control over how the `Authorization` header is built.

Sorry we don't have all the fancy bells and whistles that the C# version has, but hopefully this points you in the right direction.

**Comment by riccardobecker** [comment]
Hugues,

thanks for your reply. I made some changes to the servicebusservice.py so that I can also use already generated SAS keys for devices. Thanks for your directions.

regards

**Comment by bearrito** [comment]
Is it possible to share those changes?

Does the client still not support used generated  as described here - https://azure.microsoft.com/en-us/documentation/articles/service-bus-shared-access-signature-authentication/

**Comment by lmazuel** [comment]
Here is the sample to connect using SAS:
http://azure-sdk-for-python.readthedocs.io/en/latest/servicebus.html#shared-access-signature-authentication

Tell me if you need something else!

**Comment by bearrito** [comment]
I had the same question as the OP which was using generated SAS tokens e.g. use the key from the portal to generate something like `SharedAccessSignature sig=<signature-string>&se=<expiry>&skn=<keyName>&sr=<URL-encoded-resourceURI>
`

It looks not supported in this version of the client?

**Comment by lmazuel** [comment]
@bearrito ok, understood.
@riccardobecker do you still have the code you used to fix your problem?

**Comment by lmazuel** [comment]
@bearrito I just tested, using SAS to authenticate with policy to a specific queue/topic works with the existing code. Otherwise I miss something?

**Comment by bearrito** [comment]
So our setup is what is suggested here https://azure.microsoft.com/en-us/documentation/articles/storage-dotnet-shared-access-signature-part-1/

`2. A lightweight service authenticates the client as needed and then generates a SAS. Once the client receives the SAS, they can access storage account resources directly with the permissions defined by the SAS and for the interval allowed by the SAS. The SAS mitigates the need for routing all data through the front-end proxy service.`

I created the below subclass which does what I need.

https://gist.github.com/bearrito/4404431a27282644030946faf9f7c4f0


I also had issues with the way the resource is constructed I found that the token server had to use https for the protocol and we had to append messages to the resource

`
